/**:
  ros__parameters:
    option:
      # publish
      is_publishing_debug_visualization_marker: true
      is_publishing_clearance_map: false # publish clearance map as nav_msgs::OccupancyGrid
      is_publishing_object_clearance_map: false # publish clearance map as nav_msgs::OccupancyGrid
      is_publishing_area_with_objects: false # publish occupancy map as nav_msgs::OccupancyGrid

      is_stopping_if_outside_drivable_area: true # stop if the ego's footprint will be outside the drivable area

      # show
      is_showing_debug_info: false
      is_showing_calculation_time: false

      # other
      enable_avoidance: false # enable avoidance function
      enable_pre_smoothing: true # enable EB
      skip_optimization: false # skip MPT and EB
      reset_prev_optimization: false

    common:
      # sampling
      num_sampling_points: 100 # number of optimizing points

      # trajectory total/fixing length
      trajectory_length: 300.0 # total trajectory length[m]

      forward_fixing_min_distance: 1.0 # number of fixing points around ego vehicle [m]
      forward_fixing_min_time: 0.5 # forward fixing time with current velocity [s]

      backward_fixing_distance: 5.0 # backward fixing length from base_link [m]
      delta_arc_length_for_trajectory: 0.1 # delta arc length for trajectory [m]

      delta_dist_threshold_for_closest_point: 3.0 # delta dist thres for closest point [m]
      delta_yaw_threshold_for_closest_point: 1.046 #M_PI/3.0, delta yaw thres for closest point
      delta_yaw_threshold_for_straight: 0.02 # delta dist thres for straight point

      num_fix_points_for_extending: 50 # number of fixing points when extending
      max_dist_for_extending_end_point: 0.0001 # minimum delta dist thres for extending last point [m]

      enable_clipping_fixed_traj: false
      non_fixed_trajectory_length: 5.0 # length of the trajectory merging optimized mpt trajectory to original(not optimized) trajectory

    object:  # avoiding object
      max_avoiding_objects_velocity_ms: 0.5 # maximum velocity for avoiding objects [m/s]
      max_avoiding_ego_velocity_ms: 6.0 # maximum ego velocity when avoiding objects [m/s]

      avoiding_object_type:
        unknown: true
        car: true
        truck: true
        bus: true
        bicycle: true
        motorbike: true
        pedestrian: true
        animal: true

    # mpt param
    mpt:
      option:
        steer_limit_constraint: true
        fix_points_around_ego: true
        plan_from_ego: true
        max_plan_from_ego_length: 10.0
        visualize_sampling_num: 1
        enable_manual_warm_start: true
        enable_warm_start: true # false
        is_fixed_point_single: false

      common:
        num_curvature_sampling_points: 5 # number of sampling points when calculating curvature
        delta_arc_length_for_mpt_points: 0.5 # delta arc length when generating mpt [m]

      # kinematics:
        # If this parameter is commented out, the parameter is set as below by default.
        # The logic could be `optimization_center_offset = vehicle_info.wheel_base * 0.8`
        # The 0.8 scale is adopted as it performed the best.
        # optimization_center_offset: 2.3 # optimization center offset from base link

    # replanning & trimming trajectory param outside algorithm
    replan:
      max_path_shape_change_dist: 0.3 # threshold of path shape change from behavior path [m]
      max_ego_moving_dist_for_replan: 3.0 # threshold of ego's moving distance for replan [m]
      max_delta_time_sec_for_replan: 1.0 # threshold of delta time for replan [second]

    # advanced parameters to improve performance as much as possible
    advanced:
      eb:
        common:
          num_joint_buffer_points: 3 # number of joint buffer points
          num_offset_for_begin_idx: 2 # number of shifting points for beginning non-fix idx
          delta_arc_length_for_eb: 0.6 # 1.0 # delta arc length when optimizing[m] When the value is around 1.0, overshoot at the corner happens.
          num_sampling_points_for_eb: 95 # number of optimizing points # must be longer than mpt points

        clearance:
          clearance_for_straight_line: 0.05 # minimum optimizing range around straight points
          clearance_for_joint: 0.1 # minimum optimizing range around joint points
          clearance_for_only_smoothing: 0.1 # minimum optimizing range when applying only smoothing

        qp:
          max_iteration: 10000 # max iteration when solving QP
          eps_abs: 1.0e-8 # eps abs when solving OSQP
          eps_rel: 1.0e-10 # eps rel when solving OSQP

      mpt:
        bounds_search_widths: [0.45, 0.15, 0.05, 0.01]

        clearance:  # clearance(distance) between vehicle and roads/objects when generating trajectory
          hard_clearance_from_road: 0.0 # clearance from road boundary[m]
          soft_clearance_from_road: 0.1 # clearance from road boundary[m]
          soft_second_clearance_from_road: 1.0 # clearance from road boundary[m]
          clearance_from_object: 1.0 # clearance from object[m]
          extra_desired_clearance_from_road: 0.0 # extra desired clearance from road

        weight:
          soft_avoidance_weight: 1000.0 # slack weight for lateral error around the middle point
          soft_second_avoidance_weight: 100.0 # slack weight for lateral error around the middle point

          lat_error_weight: 100.0 # weight for lateral error
          yaw_error_weight: 0.0 # weight for yaw error
          yaw_error_rate_weight: 0.0 # weight for yaw error rate
          steer_input_weight: 10.0 # weight for steering input
          steer_rate_weight: 10.0 # weight for steering rate

          obstacle_avoid_lat_error_weight: 3.0 # weight for lateral error
          obstacle_avoid_yaw_error_weight: 0.0 # weight for yaw error
          obstacle_avoid_steer_input_weight: 1000.0 # weight for yaw error
          near_objects_length: 30.0 # weight for yaw error

          terminal_lat_error_weight: 100.0 # weight for lateral error at terminal point
          terminal_yaw_error_weight: 100.0 # weight for yaw error at terminal point
          terminal_path_lat_error_weight: 1000.0 # weight for lateral error at path end point
          terminal_path_yaw_error_weight: 1000.0 # weight for yaw error at path end point

        # check if planned trajectory is outside drivable area
        collision_free_constraints:
          option:
            l_inf_norm: true
            soft_constraint: true
            hard_constraint: false
            # two_step_soft_constraint: false

          vehicle_circles:
            method: "rear_drive"

            uniform_circle:
              num: 3
              radius_ratio: 0.8

            rear_drive:
              num_for_calculation: 3
              front_radius_ratio: 1.0
              rear_radius_ratio: 1.0

            bicycle_model:
              num_for_calculation: 3
              front_radius_ratio: 1.0
              rear_radius_ratio: 1.0
